pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go

```diff
@@ -329,7 +366,38 @@ func activateTenant(ctx context.Context, execCtx interface{}, newTenantID roachp
 // TODO(adityamaru): Add ClearRange logic once we have introduced
 // synchronization between the flow tearing down and the job transitioning to a
 // failed/canceled state.
-func (s *streamIngestionResumer) OnFailOrCancel(_ context.Context, _ interface{}) error {
+func (s *streamIngestionResumer) OnFailOrCancel(ctx context.Context, execCtx interface{}) error {
+	jobExecCtx := execCtx.(sql.JobExecContext)
+	p := s.job.Payload()
+	details := s.job.Details().(jobspb.StreamIngestionDetails)
+
+	// Cancel the producer job on best effort. The source job's protected timestamp is no
+	// longer needed as this ingestion job is in 'reverting' status and we won't resume
+	// ingestion anymore.
+	s.cancelProducerJob(ctx,
+		streaming.StreamID(details.StreamID), streamingccl.StreamAddress(details.StreamAddress))
+
+	tr, err := sql.GetTenantRecord(ctx, jobExecCtx.ExecCfg(), jobExecCtx.Txn(), details.NewTenantID.ToUint64())
+	if err != nil {
+		return err
+	}
+
+	if tr.State == descpb.TenantInfo_ACTIVE {
+		return errors.Errorf("error in cancel stream ingestion: cannot delete an active tenant %s",
+			details.NewTenantID)
+	}
+
+	// TODO(casper): deal with very large tenant.
+	tenantInfo := descpb.TenantInfo{
+		ID:    p.GetStreamIngestion().NewTenantID.ToUint64(),
+		State: descpb.TenantInfo_DROP,
+	}
+	// GC the tenant's data as well as the system tenant's knowledge about this tenant.
+	if err = sql.GCTenantSync(ctx, jobExecCtx.ExecCfg(), &tenantInfo); err != nil {
+		log.Warningf(ctx,
+			"encountered error when clearing tenant key ranges for ingestion job %d", s.job.ID())
+		return err
+	}
```

Comment:
Per recent discussions, I think for now we may want to avoid doing this altogether and have a second command: either just `SELECT crdb_internal.destroy_tenant(ID, true)` or a `DROP TENANT` command that we create.